java/ql/lib/semmle/code/java/regex/regex.qll

```diff
@@ -0,0 +1,933 @@
+        index = 1
+        or
+        this.quoteDelimiter(index - 1, _) = false
+      ) and
+      inner_start = start + 2 and
+      inner_end = end - 2 and
+      inner_end > inner_start and
+      this.quoteDelimiter(inner_end) = false and
+      not exists(int mid |
+        this.quoteDelimiter(mid) = false and mid in [inner_start .. inner_end - 1]
+      )
+    )
+  }
+
+  /**
+   * A control sequence, `\cx`
+   * `x` may be any ascii character including special characters.
+   */
+  predicate controlEscape(int start, int end) {
+    this.escapingChar(start) and
+    this.getChar(start + 1) = "c" and
+    end = start + 3
+  }
+
+  /** Gets the text of this regex */
+  string getText() { result = this.(StringLiteral).getValue() }
+
+  string getChar(int i) { result = this.getText().charAt(i) }
+
+  string nonEscapedCharAt(int i) {
+    result = this.getText().charAt(i) and
+    not exists(int x, int y | this.escapedCharacter(x, y) and i in [x .. y - 1]) and
+    not exists(int x, int y | this.quote(x, y) and i in [x .. y - 1]) and
+    not exists(int x, int y | this.controlEscape(x, y) and i in [x .. y - 1])
+  }
+
+  private predicate isOptionDivider(int i) { this.nonEscapedCharAt(i) = "|" }
+
+  private predicate isGroupEnd(int i) { this.nonEscapedCharAt(i) = ")" and not this.inCharSet(i) }
+
+  private predicate isGroupStart(int i) { this.nonEscapedCharAt(i) = "(" and not this.inCharSet(i) }
+
+  predicate failedToParse(int i) {
+    exists(this.getChar(i)) and
+    not exists(int start, int end |
+      this.topLevel(start, end) and
+      start <= i and
+      end > i
+    )
+  }
+
+  /** An escape sequence that includes braces, such as named characters (\N{degree sign}), named classes (\p{Lower}), or hex values (\x{h..h}) */
+  private predicate escapedBraces(int start, int end) {
+    this.escapingChar(start) and
+    this.getChar(start + 1) = ["N", "p", "P", "x"] and
+    this.getChar(start + 2) = "{" and
+    this.getChar(end - 1) = "}" and
+    end > start and
+    not exists(int i | start + 2 < i and i < end - 1 | this.getChar(i) = "}")
+  }
+
+  /**
+   * Holds if an escaped character is found between `start` and `end`.
+   * Escaped characters include hex values, octal values and named escapes,
+   * but excludes backreferences.
+   */
+  predicate escapedCharacter(int start, int end) {
+    this.escapingChar(start) and
+    not this.backreference(start, _) and
+    (
+      // hex value \xhh
+      this.getChar(start + 1) = "x" and
+      this.getChar(start + 2) != "{" and
+      end = start + 4
+      or
+      // octal value \0o, \0oo, or \0ooo. Max of 0377.
+      this.getChar(start + 1) = "0" and
+      end in [start + 3 .. start + 5] and
+      forall(int i | i in [start + 1 .. end - 1] | this.isOctal(i)) and
+      (end = start + 5 implies this.getChar(start + 2) <= "3") and
+      not (
+        end < start + 5 and
+        this.isOctal(end) and
+        (end = start + 4 implies this.getChar(start + 2) <= "3")
+      )
```

Comment:
This seems a bit complicated, how about just using some if-then-else to enumerate the 3 cases.
```suggestion
      // octal value \0o, \0oo, or \0ooo. Max of 0377.
      this.getChar(start + 1) = "0" and
      this.isOctal(start + 2) and
      if this.isOctal(start + 3)
      then
        if this.isOctal(start + 4) and this.getChar(start + 2) in ["0", "1", "2", "3"]
        then
          end = start + 5
        else
          end = start + 4
      else
        end = start + 3
```